src/platform/backends/lxd/lxd_vm_image_vault.cpp

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright (C) 2020 Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "lxd_vm_image_vault.h"
+
+#include <multipass/exceptions/aborted_download_exception.h>
+#include <multipass/format.h>
+#include <multipass/logging/log.h>
+#include <multipass/rpc/multipass.grpc.pb.h>
+#include <multipass/url_downloader.h>
+#include <multipass/vm_image.h>
+#include <multipass/vm_image_host.h>
+
+#include <QRegularExpression>
+
+#include <chrono>
+#include <thread>
+
+namespace mp = multipass;
+namespace mpl = multipass::logging;
+
+using namespace std::literals::chrono_literals;
+
+namespace
+{
+constexpr auto category = "image vault";
+
+auto parse_percent_as_int(const QString& progress_string)
+{
+    QRegularExpression re{"\\s(\\d{1,3})%"};
+
+    QRegularExpressionMatch match = re.match(progress_string);
+
+    if (match.hasMatch())
+    {
+        return match.captured(1).toInt();
+    }
+
+    return -1;
+}
+} // namespace
+
+mp::LXDVMImageVault::LXDVMImageVault(std::vector<VMImageHost*> image_hosts, const QUrl& base_url)
+    : image_hosts{image_hosts}, base_url{base_url}, manager{std::make_unique<mp::NetworkAccessManager>()}
```

Comment:
{C,Sh}ould the manager not be shared?

Reply:
I think the only way to make that work and still be generic enough for all image vault & backend driver cases would be to refactor/expand `URLDownloader` to use the new `NetworkManager` class.

Reply:
I just mean that, ideally, we'd have a shared manager for everything LXD… Not saying that URLDownloader should, too.

My rationale being that we could then set it up once (especially with SSL and authentication) and share it through all LXD components. Also saving memories.

Reply:
The problem I'm seeing here is that `DaemonConfig` is responsible for setting up the image vault and setting up the backend driver and does that as two separate steps that are only coupled by `URLDownloader`.  In order for `LXDVMImageVault` and `LXDVirtualMachineFactory` to share a `NetworkAccessManager` object would be to somehow relate those two, but do it in a way that `DaemonConfig` does not have to account for backend and image host differences.  I think we want `DaemonConfig` to be a generic as possible and we achieve that via the `platform` calls, but still, those platform calls do not couple the image vault and the backed driver.

If we don't go the `URLDownloader` way, then we'd need to refactor the `VirtualMachineFactory` to accept the `VMImageVault` object (at least a pointer to it) and then each factory and image vault can determine their relationship to one another.  Some would just ignore each other like now and we'd basically be forcing it hold at least a pointer to the vault and others like LXD would have a fairly coupled relationship like sharing the `NetworkAcessManager`, etc.

I'll think about it some more, but I don't think it will be trivial.

Reply:
Bear in mind none of the comments I made here are my requirement, or stuff to be fixed in this PR. It's just driving my thinking about the architecture, since this is pushing what we had in place.

Reply:
Sure, and I totally agree with the idea of being able to couple the image vault and vm factory when needed.